### PR TITLE
Support dtype as string or class in astype

### LINKED ISSUE #2028: `astype` does not accept dtype specified as `str`
I believe we are now support strings (such as 'float32') as dtype in most other places, but `astype` does not work with this.

### PR DESCRIPTION
**Pull request by jl-wynen**
Fixes #2028

### CODE REVIEW

File: dataset/to_unit.cpp [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+#include "scipp/dataset/to_unit.h"
+#include "scipp/variable/to_unit.h"
+
+namespace scipp::dataset {
+
+DataArray to_unit(const DataArray &array, const units::Unit &unit,
+                  const CopyPolicy copy) {
+  return DataArray(to_unit(array.data(), unit, copy), array.coords(),
+                   array.masks(), array.attrs(), array.name());
```
Comment (SimonHeybrock) [inline dataset/to_unit.cpp:15]: Masks should also be affected by the copy policy, compare other operations that modify data, e.g., https://github.com/scipp/scipp/blob/94bb70486d9a9a7296393f9fab0089ec771d62d4/templates/dataset_unary.cpp.in#L10-L12

File: dataset/to_unit.cpp [hunk 2]
```diff
@@ -0,0 +1,18 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2021 Scipp contributors (https://github.com/scipp)
+/// @file
+/// @author Jan-Lukas Wynen
+#include <units/units.hpp>
```
Comment (SimonHeybrock) [inline dataset/to_unit.cpp:5]: Unused?

File: python/bind_operators.h [hunk 3]
```diff
@@ -34,21 +37,26 @@ void bind_common_operators(pybind11::class_<T, Ignored...> &c) {
...
+      [](const T &self, const py::object &type, const bool copy) {
+        const auto [scipp_dtype, dtype_unit] =
+            cast_dtype_and_unit(type, std::nullopt);
+        const auto unit =
+            dtype_unit == scipp::units::one ? self.unit() : dtype_unit;
+        [[maybe_unused]] py::gil_scoped_release release;
+        return to_unit(astype(self, scipp_dtype,
+                              copy ? scipp::CopyPolicy::Always
+                                   : scipp::CopyPolicy::TryAvoid),
+                       unit, scipp::CopyPolicy::TryAvoid);
```
Comment (SimonHeybrock) [inline python/bind_operators.h:57]: Is this for datetime? I think we should *not* support this. Unlike numpy, scipp tries to separate dtype and unit, and I think we should do the same here.

**Review by SimonHeybrock — CHANGES_REQUESTED** [review]

File: dataset/to_unit.cpp (hunk 1 above)
Comment (jl-wynen) [reply dataset/to_unit.cpp:15]: Would this depend on whether the data is copied? Or would it always not copy if `copy == CopyPolicy::TryAvoid`?

Comment (SimonHeybrock) [reply dataset/to_unit.cpp:15]: I think it should depend on whether data is copied or not.

File: dataset/test/data_array_test.cpp [hunk 4]
```diff
@@ -91,12 +93,70 @@ TEST_F(DataArrayTest, shadow_attr) {
...
+  EXPECT_EQ(converted.masks(), original.masks());
+
+  EXPECT_TRUE(converted.coords()[Dim::X].is_same(original.coords()[Dim::X]));
+  if (expect_copy) {
+    EXPECT_FALSE(converted.data().is_same(original.data()));
+    EXPECT_FALSE(converted.masks()["m"].is_same(original.masks()["m"]));
+  } else {
+    EXPECT_TRUE(converted.data().is_same(original.data()));
+    EXPECT_TRUE(converted.masks()["m"].is_same(original.masks()["m"]));
+  }
```
Comment (SimonHeybrock) [inline dataset/test/data_array_test.cpp:112]:
```suggestion
  EXPECT_EQ(converted.data().is_same(original.data()), !expect_copy);
  EXPECT_EQ(converted.masks()["m"].is_same(original.masks()["m"]), !expect_copy);
```

File: dataset/test/data_array_test.cpp [hunk 5]
```diff
@@ -91,12 +93,70 @@ TEST_F(DataArrayTest, shadow_attr) {
...
+  DataArray a(makeVariable<double>(Dims{Dim::X}, Shape{3},
+                                   Values{1.0, 2.0, 3.0}, units::m),
+              {{Dim::X, makeVariable<int>(Dims{Dim::X}, Shape{3},
+                                          Values{4, 5, 6}, units::s)}},
+              {{"m", makeVariable<bool>(Dims{Dim::X}, Shape{3},
+                                        Values{true, false, true})}});
+  check_to_unit(a, units::mm, CopyPolicy::TryAvoid, true);
+  check_to_unit(a, units::mm, CopyPolicy::Always, true);
+  check_to_unit(a, units::m, CopyPolicy::TryAvoid, false);
+  check_to_unit(a, units::m, CopyPolicy::Always, true);
```
Comment (SimonHeybrock) [inline dataset/test/data_array_test.cpp:159]: Can we move these tests to another file? I think we should have mainly tests of the `class DataArray` here.

**Review by SimonHeybrock — APPROVED** [review]

**Review by SimonHeybrock — APPROVED** [review]

**Pull request merged**